### ISSUE TITLE
Loader drops pin-level info: net_nodes only contains macro indices

### ISSUE THREAD

**Issue by willpartcl** [partcleda/macro-place-challenge-2026#70]
## Problem

`Benchmark.net_nodes` exposes only per-macro connectivity, not per-pin. Teams building gradient-based placers that need pin-level HPWL (to match the evaluator's wirelength formula) cannot reconstruct it from the `Benchmark` object alone — they have to re-parse `netlist.pb.txt` themselves.

Reported by Tobias Franks (ByteDancer / Incremental CD) on the Contact mailing list.

## Where it happens

`macro_place/loader.py:147-156`:

```python
for driver, sinks in plc.nets.items():
    nodes_in_net = set()
    for pin_name in [driver] + sinks:
        # Pin names are "MACRO/PIN" for macro pins or just "PORT" for ports
        parent = pin_name.split("/")[0]   # drops the /PIN part
        if parent in name_to_bench:
            nodes_in_net.add(name_to_bench[parent])   # set dedups
    if nodes_in_net:
        net_nodes.append(torch.tensor(sorted(nodes_in_net), dtype=torch.long))
```

Two lossy steps:
1. `pin_name.split("/")[0]` discards the pin name (`a22/IP1` → `a22`).
2. `nodes_in_net = set()` collapses multiple pins on the same macro into one entry.

## Why this matters

The evaluator (`plc_client_os`) computes wirelength per pin using the actual pin position = `macro_center + pin_offset`. Placers wanting to match this in a differentiable loss need to know, for each net, *which* pin on each connected macro is the endpoint.

`macro_pin_offsets` (added in cab647b) exposes per-macro pin geometry, but without a net→pin mapping it's unusable for pin-level HPWL. Competitors have three options today:
- Approximate with the macro centerpoint (worse gradients → worse placements)
- Sum all pin-to-pin distances on each net assuming all pins of every connected macro participate (overestimates)
- Re-parse `netlist.pb.txt` to recover the mapping (extra work, what Tobias did)

## Proposed fix

Add `net_pin_nodes: List[Tensor]` to `Benchmark`. Each entry shape `[pins_in_net, 2]` with columns `(owner_idx, pin_slot)`:
- Hard macros: `pin_slot` indexes into `macro_pin_offsets[owner]`
- Soft macros / I/O ports: `pin_slot = 0` (no per-pin offsets stored)

In the loader, build a `pin_name → (macro_name, slot)` map while extracting `macro_pin_offsets`, then use it in the net-parsing loop. Keep `net_nodes` as-is (backwards compat); dedup+sort of `net_pin_nodes[:, 0]` equals `net_nodes` for every net.

`Benchmark.save` / `Benchmark.load` round-trip the new field; old `.pt` files load with an empty list (backwards-compatible). Regenerate via `scripts/convert_ibm_benchmarks.py` to populate.

## Impact on the leaderboard

Eval scores are unaffected — everyone gets the exact proxy via `compute_proxy_cost(placement, benchmark, plc)` regardless of internal loss. What's affected: gradient-based placers (ByteDancer, Pragnay, W3's hybrid, TAISPlAce, BakaBobo, Convex Opt, Waterloo kid) that are currently either approximating or parsing netlists themselves.
